zkevm-circuits/src/keccak_circuit/keccak_packed.rs

```diff
@@ -0,0 +1,1770 @@
+                        for ((part_a, part_b), part_c) in os_parts[i][j]
+                            .iter()
+                            .zip(os_parts[(i + 1) % 5][j].iter())
+                            .zip(os_parts[(i + 2) % 5][j].iter())
+                        {
+                            //println!("part size: {}", part.num_bits);
+                            let expr = pack_bit(3, part_size_base) + part_b.expr.clone()
+                                - 2.expr() * part_a.expr.clone()
+                                - part_c.expr.clone();
+                            fat.push(Part {
+                                num_bits: part_size_base,
+                                expr: expr.clone(),
+                                parts: vec![expr.clone()],
+                            });
+                        }
+                    } else {
+                        let input = pack_bit(3, 64) + s[(i + 1) % 5][j].clone()
+                            - 2.expr() * s[i][j].clone()
+                            - s[(i + 2) % 5][j].clone();
+                        fat = split(
+                            meta,
+                            &mut cell_values,
+                            &mut cb,
+                            input,
+                            0,
+                            part_size_base,
+                            false,
+                        );
+                    }
+                    os[i][j] = decode(transform(
+                        "chi base",
+                        meta,
+                        &mut cell_values,
+                        &mut lookup_counter,
+                        fat.clone(),
+                        chi_base_table,
+                    ));
+                    cb.require_equal("next row check", os[i][j].clone(), s_next[i][j].clone());
+                }
+            }
+        }
+        println!("- Post chi:");
+        println!("Lookups: {}", lookup_counter);
+        println!("Columns: {}", cell_values.len());
+        println!("num_parts_pre: {}", num_parts_pre);
+        println!("num_parts_post: {}", num_parts_post);
+        total_lookup_counter += lookup_counter;
+
+        // Unpack a single word into bytes (for the squeeze)
+        // Potential optimization: could potentially do multiple bytes per lookup
+        let packed = split(
+            meta,
+            &mut cell_values,
+            &mut cb,
+            squeeze_from.clone(),
+            0,
+            8,
+            false,
+        );
+        transform(
+            "squeeze unpack",
+            meta,
+            &mut cell_values,
+            &mut lookup_counter,
+            packed,
+            pack_unpack_table
+                .into_iter()
+                .rev()
+                .collect::<Vec<_>>()
+                .try_into()
+                .unwrap(),
+        );
+
+        meta.create_gate("input checks", |meta| {
+            let mut cb = BaseConstraintBuilder::new(5);
```

Comment:
Sometimes cleaner to have a constraint builder per gate? But that doesn't always work well it kind of mixes both ways. To be honest though it's because it's currently not very well thought out to make the most sense.

Reply:
I'm not sure either. But seemed a bit odd to me to have multiple CB's within the same context.